File: contrast.py
"""
contrast: ranges that most separate two groups of rows.
X ranges come from the parent table; one XY per column counts
which group lands in each range. Contrast is just counting.
(c) 2026 Tim Menzies <timm@ieee.org> MIT license
"""
import sys; sys.dont_write_bytecode = True
from tbl import *

def score(z): return z.score

def discretize(d0, d1, ordered=False): # two count dicts
  "Yield spans scored b^2/(b+r); b,r = frequency in d0,d1."
  n0 = sum(d0.values()) + 1/BIG
  n1 = sum(d1.values()) + 1/BIG
  ks = sorted(set(d0) | set(d1))
  B, R, m = [0], [0], len(ks)
  for k in ks:
    B += [B[-1] + d0.get(k, 0)/n0]
    R += [R[-1] + d1.get(k, 0)/n1]
  for i in range(m):
    for j in (range(i + 1, m + 1) if ordered else [i + 1]):
      if (i, j) != (0, m): # whole column selects nothing
        yield _span(ks[i], ks[j-1], B[j]-B[i], R[j]-R[i],
                    i == 0, j == m)

def _span(lo, hi, b, r, first, last): # the scorer: b^2/(b+r)
  return o(lo=lo, hi=hi, first=first, last=last,
           score=b*b / (b + r + 1/BIG))

def contrasts(tbl, rows0, rows1): # best range splitting groups
  cs = ranges(tbl)
  def go(at):
    xy = XY(cs.get(at))
    for y, rows in ((0, rows0), (1, rows1)):
      for r in rows: add(xy, (r[at], y))
    d0 = {b: d.get(0, 0) for b, d in xy.seen.items()}
    d1 = {b: d.get(1, 0) for b, d in xy.seen.items()}
    z = max((z for d in (discretize(d0, d1, at in cs),
                         discretize(d1, d0, at in cs))
             for z in d), key=score, default=None)
    if z:
      z.at, name = at, tbl.cols.names[at]
      if at not in cs:
        z.txt = f"{name} == {z.lo}"
      else:
        z.lo = -BIG if z.first else cs[at][z.lo]
        z.hi =  BIG if z.last  else cs[at][z.hi + 1]
        z.txt = (f"{name} <= {z.hi:.3g}" if z.lo == -BIG else
                 f"{name} >= {z.lo:.3g}" if z.hi == BIG else
                 f"{name} in {z.lo:.3g}..{z.hi:.3g}")
    return z
  return max((z for at in tbl.cols.x if (z := go(at))),
             key=score, default=None)

def selects(cut, row): # does row fall inside a contrast range?
  v = row[cut.at]
  return v != "?" and cut.lo <= v <= cut.hi


# ---------------------------------------------------------------
def test_discretize():
  "two bin-count dicts; best span covers the first"
  d0 = {0: 30, 1: 50, 2: 20}
  d1 = {6: 40, 7: 40, 8: 20}
  z = max(discretize(d0, d1, ordered=True), key=score)
  assert z.score > 0.9 and z.lo == 0 and z.hi == 2
  print("d0 in bins 0-2, d1 in bins 6-8:",
        "span %s..%s score %.2f" % (z.lo, z.hi, z.score))

def test_contrasts():
  "name the 50:50 halves of fastmap's top-level split"
  import unsuper
  t = Tbl(csv(the.file))
  x = lambda r1, r2: unsuper.distx(t, r1, r2)
  _, rows = unsuper.fastmap(t.rows[:], x)
  n = len(rows) // 2
  r0, r1 = rows[:n], rows[n:]
  z = contrasts(t, r0, r1)
  p0 = sum(selects(z, r) for r in r0) / len(r0)
  p1 = sum(selects(z, r) for r in r1) / len(r1)
  print("best: %s; selects %.0f%% vs %.0f%%" %
        (z.txt, 100*p0, 100*p1))
  assert abs(p0 - p1) > 0.2

def test_halve():
  "one line: best range splitting fastmap's two halves"
  import unsuper
  t = Tbl(csv(the.file))
  x = lambda r1, r2: unsuper.distx(t, r1, r2)
  _, rows = unsuper.fastmap(t.rows[:], x)
  n = len(rows) // 2
  z = contrasts(t, rows[:n], rows[n:])
  print("%-45s %.2f %s" % (the.file.split("/")[-1],
                           z.score if z else 0,
                           z.txt if z else "no split"))

if __name__ == "__main__": main(globals())

File: contrast.lua
-- contrast: ranges that most separate two sets of rows.
-- Numeric ranges come from the.bins divisions of -3sd..+3sd.
-- (c) 2026 Tim Menzies <timm@ieee.org> MIT license
local _ENV = setmetatable({}, {__index=require"tbl"})

function discretize(best,rest,out,      f) -- spans, both ways
  f = isNum(best) and _discretizeNum or _discretizeSym
  f(best, rest, out)
  f(rest, best, out) end

function _span(lo,hi,b,r,out) -- discretize's scorer: b^2/(b+r)
  out{lo=lo, hi=hi, score=b*b / (b + r + 1/BIG)} end

function _discretizeSym(best,rest,out,      n0,n1,keys)
  -- one span per symbol
  n0,n1,keys = 1/BIG, 1/BIG, {}
  for k,n in pairs(best) do n0 = n0 + n; keys[k] = true end
  for k,n in pairs(rest) do n1 = n1 + n; keys[k] = true end
  for k in pairs(keys) do
    _span(k, k, (best[k] or 0)/n0, (rest[k] or 0)/n1, out) end
end

function _discretizeNum(best,rest,out, -- the.bins cut spans
                        n,lo,hi,cuts,B,R)
  n  = the.bins
  lo = math.min(best[2]-3*sd(best), rest[2]-3*sd(rest))
  hi = math.max(best[2]+3*sd(best), rest[2]+3*sd(rest))
  cuts,B,R = {},{},{}
  for k = 0,n do
    cuts[k] = lo + k*(hi - lo)/n
    B[k] = norm(best, cuts[k])
    R[k] = norm(rest, cuts[k]) end
  cuts[0], cuts[n] = -BIG, BIG -- extreme spans are open-ended
  for i = 0,n-1 do
    for j = i+1,n do
      if not (i == 0 and j == n) then -- whole column useless
        _span(cuts[i],cuts[j], B[j]-B[i], R[j]-R[i], out) end
    end end end

function contrasts(tbl0,tbl1,      score,all,one,z)
  -- best span over all x columns
  score = function(z1) return z1.score end
  all = most(score)
  for at,_ in pairs(tbl0.cols.x) do
    one = most(score)
    discretize(tbl0.cols.all[at], tbl1.cols.all[at], one)
    z = one()
    if z then
      z.at, z.txt = at, cutTxt(tbl0, at, z)
      all(z) end end
  return all() end

function cutTxt(tbl,at,z,      name) -- span --> readable text
  name = tbl.cols.names[at]
  if not isNum(tbl.cols.all[at]) then
    return ("%s == %s"):format(name, z.lo) end
  if z.lo == -BIG then
    return ("%s <= %.3g"):format(name, z.hi) end
  if z.hi == BIG then
    return ("%s >= %.3g"):format(name, z.lo) end
  return ("%s in %.3g..%.3g"):format(name, z.lo, z.hi) end

function selects(cut,row,      v) -- row inside the range?
  v = row[cut.at]
  return v ~= "?" and cut.lo <= v and v <= cut.hi end


-- --------------------------------------------------------------
function test_discretize() -- separated Nums; span covers one
  local a,b = Num(), Num()
  for _ = 1,100 do add(a, rand()); add(b, rand() + 2) end
  local out = most(function(z) return z.score end)
  discretize(a, b, out)
  local z = out()
  assert(z.score > 0.5 and (z.hi < 2 or z.lo > 1))
  print(("a=N(.5,.3) b=N(2.5,.3): %.2f..%.2f score %.2f")
        :format(math.max(z.lo,-9), math.min(z.hi,9), z.score))
end

function test_contrasts() -- name fastmap's 50:50 top split
  local u = require"unsuper"
  local t = Tbl(csv(the.file))
  local x = function(r1,r2) return u.distx(t,r1,r2) end
  local _,rows = u.fastmap(t.rows, x)
  local half, r0, r1 = #rows // 2, {}, {}
  for i,r in ipairs(rows) do
    if i <= half then r0[#r0+1] = r else r1[#r1+1] = r end end
  local t0, t1 = clone(t,r0), clone(t,r1)
  local z = contrasts(t0, t1)
  local p0, p1 = 0, 0
  for _,r in ipairs(t0.rows) do
    p0 = p0 + (selects(z,r) and 1 or 0) end
  for _,r in ipairs(t1.rows) do
    p1 = p1 + (selects(z,r) and 1 or 0) end
  p0, p1 = p0/#t0.rows, p1/#t1.rows
  print(("best: %s; selects %.0f%% vs %.0f%%")
        :format(z.txt, 100*p0, 100*p1))
  assert(math.abs(p0 - p1) > 0.2) end

function test_halve() -- one line: best split of two halves
  local u = require"unsuper"
  local t = Tbl(csv(the.file))
  local x = function(r1,r2) return u.distx(t,r1,r2) end
  local _,rows = u.fastmap(t.rows, x)
  local half, r0, r1 = #rows // 2, {}, {}
  for i,r in ipairs(rows) do
    if i <= half then r0[#r0+1] = r else r1[#r1+1] = r end end
  local z = contrasts(clone(t,r0), clone(t,r1))
  print(("%-45s %.2f %s"):format(the.file:match"([^/]+)$",
        z and z.score or 0, z and z.txt or "no split")) end

if not pcall(debug.getlocal, 4, 1) then main(_ENV) end
return _ENV
